File: Python/SUSHI.py
# 시간제한에 걸리는 코드. CPP로 동일한 알고리즘의 코드를 작성하니 통과되었다...

import sys

global foods, memo

def main():
    global foods, memo
    for _ in range(int(input())):
        foods = []
        no_of_foods, budget = map(int, input().strip().split())
        memo = [0] * (budget // 100 + 1)
        for _ in range(no_of_foods):
            foods.append(list(map(int, input().strip().split())))
        sorted(foods, key = lambda x: x[0])
            
        for i in range(0, budget // 100 + 1):
            if i == 0:
                memo[i] = 0
            else:
                for food in foods:
                    if food[0] // 100 > i:
                        break;
                    memo[i] = max(food[1] + memo[i - food[0] // 100], memo[i])
        print(memo[budget // 100])
if __name__ == '__main__':
    main()
    

File: C++/SUSHI.cpp
// 별로 어렵지 않은 문제였다.

#include <iostream>
#include <vector>
#include <cstring>

using namespace std;

int memo[21474837];
int foods[20][2];

int main(){
    int i,j,k,l;
    int no_of_foods, budget;
    for(i=0, scanf("%d", &j); i<j; i++){
        // foods.clear();
        // memset(memo, 0, sizeof(memo));
        scanf("%d %d", &no_of_foods, &budget);
        for(k=0; k<no_of_foods; k++){
            scanf("%d %d", &foods[k][0], &foods[k][1]);
        }
        for(k=0; k< budget / 100 + 1; k++){
            if(k == 0)
                memo[k] = 0;
            else{
                memo[k] = 0;
                for(l=0; l<no_of_foods; l++){
                    if(foods[l][0] / 100 <= k){
                        memo[k] = max(memo[k], foods[l][1] + memo[k - foods[l][0] / 100]);
                    }
                }
            }
        }
        std::cout << memo[budget / 100] << std::endl;
    }
    return 0;
}